Update the upgrade instructions

patiencedaur opened the pull request:
Resolves #3042 
Resolves #1807 
Resolves https://github.com/tarantool/doc/issues/836

Deploy: https://docs.d.tarantool.io/en/doc/gh-3042-caveats/book/admin/upgrades/

- Clarify the process of update in a replica set
- Move the version-specific instructions from tarantool/wiki:
  https://github.com/tarantool/tarantool/wiki/Caveats-when-upgrading-from-tarantool-1.6
  https://github.com/tarantool/tarantool/wiki/Fix-wrong-order-of-decimals-and-doubles-in-indices

patiencedaur committed:
Add note about caveats when upgrading from 1.6 to 1.10

patiencedaur committed:
Update the upgrading instructions

patiencedaur committed:
Add instructions on checking status before upgrade

patiencedaur committed:
Add version-specific upgrade instructions

Move the instructions from tarantool/wiki:

https://github.com/tarantool/tarantool/wiki/Caveats-when-upgrading-from-tarantool-1.6
https://github.com/tarantool/tarantool/wiki/Fix-wrong-order-of-decimals-and-doubles-in-indices

sergepetrenko commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,27 @@
+   Changes are propagated to other
+   nodes via the regular replication mechanism.
+
+8. Run ``box.snapshot()`` on every node in the replica set to take a snapshot of all the data.
```
I'd rather say
`to make sure replicas immediately see the upgraded database state in case of restart`

patiencedaur replied:
You mean, this is the reason we run `box.snapshot()` in this context?

sergepetrenko replied:
Yep

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +50,78 @@ How to upgrade from Tarantool 1.7 to 2.x
 
-4. Launch the updated Tarantool server using ``tarantoolctl`` or ``systemctl``.
+4. Update the Tarantool database. Put the request ``box.schema.upgrade()``
+   inside a :doc:`box.once() </reference/reference_lua/box_once>` function in your Tarantool
```
You reference this section from "updating the cluster", I think the users might get confused and think that they should run `box.schema.upgrade()` on every node that they upgrade.
Let's mention that `box.schema.upgrade()` should be issued here only when the instance is not part of the cluster.

sergepetrenko replied:
Maybe it would be better not to mention `box.schema.upgrade()` neither in "upgrading an instance" nor in "upgrading the cluster", and add a separate short paragraph "finishing the upgrade", which would describe `box.schema.upgrade()` and why it's needed.

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +50,78 @@ How to upgrade from Tarantool 1.7 to 2.x
-   If yes, back up the old application files.
+    The only way to upgrade Tarantool from 1.6 to 2.x **without downtime** is
+    taking an intermediate step by upgrading from 1.6 to 1.10 and then to 2.x.
+    This also applies to upgrading Tarantool from 1.7 and 1.9.
```
Maybe say that "the only way to upgrade Tarantool from a version prior to 1.10 (like 1.6, 1.7 and 1.9) to 2.x without downtime is to take an intermediate step..." ?

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +50,78 @@ How to upgrade from Tarantool 1.7 to 2.x
+    This also applies to upgrading Tarantool from 1.7 and 1.9.
 
-2. Stop the Tarantool server.
+    Before upgrading Tarantool from 1.x to 2.y, please read about the associated
```
But this is only about upgrading 1.6 to 1.10, isn't it?

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +50,78 @@ How to upgrade from Tarantool 1.7 to 2.x
-7. Launch the updated Tarantool server using ``tarantoolctl`` or ``systemctl``.
+       box.info.ro -- shouldn't be "orphan"
+       box.info.ro_reason -- since 2.10, shouldn't be "orphan"
+       box.info.status -- should be "running"
```
Actually we should only check that `box.info.status` is "running" on all the instances and that at least one instance has `box.info.ro` == `false`.

In case all the instances have `box.info.ro == true`, one may check `box.info.ro_reason`, starting from version 2.10, to have a clue what the issue is

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +50,78 @@ How to upgrade from Tarantool 1.7 to 2.x
+       box.info.status -- should be "running"
 
-.. _admin-upgrades_replication_cluster:
+   If ``box.info.ro``, ``box.info.ro_reason``, or ``box.info.status`` have the value ``orphan``,
```
`box.info.status` and `box.info.ro_reason`.
`box.info.ro` only holds true or false

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +50,78 @@ How to upgrade from Tarantool 1.7 to 2.x
-perform capability negotiation on handshake, and new 1.7 replication features
-are not used with 1.6 replicas. This allows upgrading clustered configurations.
+#. Make sure each replica connected to the rest of the replica set.
+   Run the following, replacing the ``id`` with the replica ID (the output of ``box.info.id``):
```
I think it would be simpler to run
`box.info.replication`, and then check the output table (which is indexed by replica ids and contains upstream, downstream for each id). Each upstream and downstream should have status = follow. The only exceptions are upstream and downstream belonging to this particular instance (they are indexed by `box.info.id`)

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +50,78 @@ How to upgrade from Tarantool 1.7 to 2.x
+   This means that the replicas are connected and there are no errors in the data flow.
 
-   a. Start Tarantool.
+   The value of the ``lag`` field should be less than 1 second.
```
I think this needs some clarification.
The value of `lag` should be sane, for example, when `box.cfg.replication_timeout` is 1 second, it's generally ok to have lag ~ 10 seconds. I guess only the user knows which values of lag are fine for him.

It's always ok to have lag <= box.cfg.replication_timeout. 
When lag is > box.cfg.replication_timeout the user has to use common sense. A lag of several hours is surely not okay (>= thousands of seconds)

Sorry if I told you otherwise in chat.

sergepetrenko commented on doc/book/admin/upgrades/1.6-1.10.rst:
```diff
@@ -0,0 +1,54 @@
+This page includes explanations and solutions to some common issues
+when upgrading the cluster from Tarantool 1.6 to 1.10.
+This will be of interest to anyone upgrading to the most recent Tarantool version,
+like 2.8.4, 2.10.0 and such, since taking an intermediate step with
```
Let's replace 2.10.0 with 2.10.1, since it's more up to date as of the time of writing

sergepetrenko commented on doc/book/admin/upgrades/1.6-1.10.rst:
```diff
@@ -0,0 +1,54 @@
+like 2.8.4, 2.10.0 and such, since taking an intermediate step with
+upgrading 1.6 -> 1.10 -> 2.x is the only way to perform the upgrade without downtime.
+
+A direct upgrade from 1.6 to 2.x is also possible, but only with downtime.
```
I can't think of a place to put it in, but I think we should mention how is a direct upgrade from 1.6 (or 1.7 or 1.9) to 2.x with downtime possible:

1. Stop all instances in the cluster
2. Upgrade Tarantool version to 2.x on every instance
3. Upgrade the corresponding instance files and applications, if needed
4. Start all the instances with tarantool 2.x
5. Execute `box.schema.upgrade()` on master
6. Execute `box.snapshot()` on every node in the cluster.

sergepetrenko reviewed (changes requested):
Thanks for working on this!
It's good our upgrade page is getting up to date and more clear.
Please, find my comments below

patiencedaur committed:
Apply some suggestions from review

* Create a 'finishing the upgrade' section with `box.schema.upgrade()`
  and other steps in it
* Clarify version numbers in the note

patiencedaur commented:
Ro_reason: https://github.com/tarantool/doc/issues/2445

patiencedaur committed:
Apply more suggestions from review

* Add instruction on direct upgrade from 1.6 to 2.x
* Clarify the purpose of ro_reason
* Clarify the appropriate `lag` value
* Provide a simpler instruction to view `upstream` and `downstream` values

patiencedaur committed:
Rephrase headers

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -47,31 +47,32 @@ instructions for individual versions :ref:`in the list below <admin-upgrades_ver
-   The value of the ``lag`` field should be less than 1 second.
+   The value of the ``lag`` field can be less or equal than ``box.cfg.replication_timeout``,
+   but it can also be moderately larger.
+   For example, if ``box.cfg.replication_timeout`` is 1 second, it's generally OK to have a lag of about 10 seconds.
```
> it's generally OK to have a lag of about 10 seconds

Let's append "... under high write load on master" there.

sergepetrenko commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +25,113 @@ How to upgrade from Tarantool 1.7 to 2.x
+    Before upgrading Tarantool from 1.6 to 2.x, please read about the associated
+    :ref:`caveats <admin-upgrades-1.6-1.10>`.
 
-Tarantool 1.7 has an incompatible :ref:`.snap <internals-snapshot>` and
-:ref:`.xlog <internals-wal>` file format: 1.6 files are
-supported during upgrade, but you won’t be able to return to 1.6 after running
-under 1.7 for a while. It also renames a few configuration parameters, but old
```
Please add this part to "caveats when upgrading 1.6 to 2.x"

Saying that tarantool 1.10 (and 2.x) have incompatible snapshot format compared to 1.6
So the user won't be able to return to 1.6 after running 1.10 or 2.x for a while

sergepetrenko commented on doc/book/admin/upgrades/1.6-1.10.rst:
```diff
@@ -0,0 +1,51 @@
+
+Once this is performed on the master, it's safe to proceed to step 8, making snapshot on every node.
+
+**Step 8:** The user might be concerned with snapshot size in 1.10 --
```
There are no steps 7 and 8 anymore, so these paragraphs should refer to step 6 (about finishing the upgrade)

sergepetrenko reviewed (changes requested):
Thanks for the fixes!
Some more comments from me

patiencedaur committed:
Apply even more suggestions from review

* Explain lag value size
* Clarify incompatibility between 1.6 and later versions
* Add schema upgrade and snapshot to instruction for 1.6 -> 1.10

sergepetrenko commented on doc/book/admin/_includes/1.6-to-2.x-condition.rst:
```diff
@@ -1,3 +1,8 @@
+Versions later that 1.6 have incompatible :ref:`.snap <internals-snapshot>` and
+:ref:`.xlog <internals-wal>` file formats: 1.6 files are
+supported during upgrade, but you won’t be able to return to 1.6 after running
+under 1.7 for a while. A few configuration parameters are also renamed.
```
"under 1.10 / 2.x"

That's up to you, but I think there's no reason to mention unsupported Tarantool versions

patiencedaur committed:
Do not mention unsupported versions

sergepetrenko reviewed (approved):
Thanks! Looks good to me.

veod32 commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +25,114 @@ How to upgrade from Tarantool 1.7 to 2.x
+   This means that the replicas are connected and there are no errors in the data flow.
 
-6. Update application files, if needed.
+   The value of the ``lag`` field can be less or equal than ``box.cfg.replication_timeout``,
```
I would make this paragpaph to be the 3d item in the numbered list -- to make the check-out preparation points more clear
1. ro/rw instances
2. upstream/downtime
3. replication lag

patiencedaur replied:
I think that it is unnecessary to divide the output of a single command into two separate steps.
Instead, I'll elaborate on the table output.

veod32 commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,21 @@
+1. Pick any replica in the replica set.
```
Should it be an RO instance? If yes, let's make this clear.

veod32 commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +25,114 @@ How to upgrade from Tarantool 1.7 to 2.x
    `download page <http://tarantool.org/download.html>`_.
 
-4. Launch the updated Tarantool server using ``tarantoolctl`` or ``systemctl``.
+After that, make sure to :ref:`finish the upgrade properly <admin-upgrades_db>`.
```
Предлагаю сделать этот пункт снова явно номером #4 в нумерованном списке и привести здесь полный текст процедуры финализации -- или инклюдом включить, или просто текстом написать.

Мотивация:

В этой маленькой инструкции из 4х шагов, мы 3 раза отсылаем пользователя в другие места выполнять другие инструкции. Это всегда неудобно, и лучше сделать по максимуму, чтобы пользователь читал инструкцию на одной странице. Если пп. 2 и 3 еще более или менее норм отсылать, то п. 4 я бы лучше привел здесь в явном виде.

К тому же мы к этой инструкции отсылаем из следущей секции (Upgrading Tarantool in a replica set with no downtime) для апдейта каждого инстанса. Получается уж больно комплексная "матрешка" из редиректов по инструкциям. Это еще один плюс к тому, чтобы сделать максимально инструкцию в одном разделе тут.

Еще один плюс к этой мотивации -- в п.1 мы стоппим инстанс. По логике, в конце мы его должны стартануть. Но в текущем описании не понятно, где. Может в п.3, а может в п.4. 
Лучше это сделать как-то явным образом, чтобы читатель видел - в начале мы стоппанули инстанс, в конце запустили. Особенно учитывая то, что к процедуре апдейта standalone инстанс мы обращаемся из шагов инструкции по "Upgrading ... with no downtime" и той инструкции проверяем коннект инстанса к остальным репликам -- логично, что для этого инстанс должен быть запущен, и лучше явно написать в процедуре апгрейда для standalone.

patiencedaur replied:
Тогда нам нужно обновить [доку по box.schema.upgrade](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_schema/upgrade/) :)

veod32 commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,21 @@
+1. Pick any replica in the replica set.
+
+2. Upgrade this replica to the new Tarantool version. See details in
+   :ref:`Upgrading a Tarantool instance <admin-upgrades_instance>`.
```
```suggestion
   :ref:`Upgrading Tarantool on a standalone instance <admin-upgrades_instance>`.
```

veod32 replied:
Учитывая, что мы отсылаем пользователя сделать процедуру "Upgrading Tarantool on a standalone instance", где инстанс
инстанс стоппится, т.е. у него есть явный даунтайм, может имеет смысл пояснить пользователю, почему мы называем процедуру в репликасете "Upgrading ... with no downtime" -- что репликасет продолжает консистентно писать и отдавать данные на оставшихся работающих репликах, а реплику, которую мы стоппанули и проапгрейдили, мы потом снова подключим к репликасету, и она "догонит" остальных  -- если я правильно понял идею этого апдегрейд with no downtime.

Может пользователю это и так очевидно, а может и нет.

veod32 commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,21 @@
+2. Upgrade this replica to the new Tarantool version. See details in
+   :ref:`Upgrading a Tarantool instance <admin-upgrades_instance>`.
+
+3. Make sure the replica connected to the rest of the replica set just fine:
```
Как писал в комменте к "Upgrading Tarantool on a standalone instance", лучше как-то явно написать, что в процедуре апгрейда одного инстанса мы его стоппим, а потом запускаем. 
В этом шаге проверяется коннект инстанса ко всем репликам в сете, инстанс явно должен быть в работающем состоянии.

veod32 commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,21 @@
+      
+   The ``status`` field in both outputs should have the value ``follow``.
+
+4. :ref:`Upgrade <admin-upgrades_instance>` all the replicas by repeating steps 1--3
```
> all the replicas

RO replicas?

veod32 commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,21 @@
+4. :ref:`Upgrade <admin-upgrades_instance>` all the replicas by repeating steps 1--3
+   until only the master keeps running the old Tarantool version.
+
+5. Make one of the updated replicas the new master.
```
>  Make one of the updated replicas the new master

How? What the procedure?

If a user needs to set `box.cfg{ read_only = false }` on this instance, let's make it clear. 
If yes:
- Can it be done on the working instance or not? It' better to make this clear as well.
- If we're setting this instance to be the new master (RW), what should we do with the former master and when? My point: if we set `box.cfg{ read_only = false }` on one of the RO instance making it the master (RW), there will be a period of time when there're 2 RW instances in the replica set -- this one (new master) and the current master until we make it RO. 

Sure we have explanation about the `read_only` option in the config reference https://www.tarantool.io/en/doc/latest/reference/configuration/#cfg-basic-read-only
but it's better to explain the main essential ToDo point here, and then to refer to the reference for other details if necessary.

veod32 commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,21 @@
+   until only the master keeps running the old Tarantool version.
+
+5. Make one of the updated replicas the new master.
+   Check that it continues following and being followed by all other replicas.
```
How? What the procedure?

If it's by checking

box.info.replication[id].upstream
box.info.replication[id].downstream

like in the step #3 above, let's make it clear by referencing to this step.

veod32 commented on doc/book/admin/_includes/upgrade_procedure.rst:
```diff
@@ -0,0 +1,21 @@
+5. Make one of the updated replicas the new master.
+   Check that it continues following and being followed by all other replicas.
+
+6. :ref:`Upgrade <admin-upgrades_instance>` the former master.
```
See comment to the step #5 above.
Let's make it more clear if we need to make the former master to be RO -- how to do this and when.

veod32 commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +25,114 @@ How to upgrade from Tarantool 1.7 to 2.x
+..  include:: ./_includes/upgrade_procedure.rst
+
+After upgrading the replica set, make sure to run ``box.schema.upgrade()`` on the new master
+as described below in the section ":ref:`Finishing the upgrade <admin-upgrades_db>`".
```
```suggestion
as described below in the section ":ref:`Finalizing the upgrade <admin-upgrades_db>`".
```

veod32 commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +25,114 @@ How to upgrade from Tarantool 1.7 to 2.x
+
+..  include:: ./_includes/upgrade_procedure.rst
+
+After upgrading the replica set, make sure to run ``box.schema.upgrade()`` on the new master
```
I would make this paragraph to be the step #7 -- it's the part of the entire procedure.

veod32 commented on doc/book/admin/upgrades.rst:
```diff
@@ -58,95 +25,114 @@ How to upgrade from Tarantool 1.7 to 2.x
---------------------------------------------------------------------------------
-Upgrading Tarantool in a replication cluster
---------------------------------------------------------------------------------
+Finally, run ``box.snapshot()`` on every node in the replica set
```
I would make this paragraph to be the step #8 -- it's the part of the entire procedure.

veod32 commented on doc/book/admin/_includes/1.6-to-2.x-condition.rst:
```diff
@@ -0,0 +1,12 @@
+it is necessary to take an intermediate step by upgrading 1.6 -> 1.10 -> 2.x.
+This is the only way to perform the upgrade without downtime.
+
+However, a direct upgrade of a replica set from 1.6 to 2.x is also possible, but only
```
> direct upgrade of a replica set from 1.6 to 2.x

Better to give a link to the description of this procedure.

patiencedaur replied:
This text is included both in the direct and stepwise instructions with the `include` directive. So I think the link won't be necessary.

veod32 commented on doc/book/admin/upgrades/1.6-1.10.rst:
```diff
@@ -0,0 +1,59 @@
+This page includes explanations and solutions to some common issues
+when upgrading a replica set from Tarantool 1.6 to 1.10.
+
+..  include:: ../_includes/1.6-to-2.x-condition.rst
```
It's not really clear why we're talking about the 1.6 > 2.x upgrade conditions in the topic for the 1.6 > 1.10 upgrade.

veod32 commented on doc/book/admin/upgrades/1.6-1.10.rst:
```diff
@@ -0,0 +1,59 @@
+
+..  include:: ../_includes/1.6-to-2.x-condition.rst
+
+Let's first reiterate the upgrade procedure for any Tarantool version:
```
May be to rephrase this intro sentence somehow? 
The previous sentence says, "However, a direct upgrade of a replica set from 1.6 to 2.x is also possible, but only with downtime."
And then we have this intro and the procedure description, but it's not clear that this is the generic procedure regardless the TT version.

veod32 commented on doc/book/admin/upgrades/1.6-1.10.rst:
```diff
@@ -0,0 +1,59 @@
+
+**Step 2:** Tarantool 1.10+ fails to recover from 1.6 xlogs, unless ``box.cfg{force_recovery = true}`` is set.
+There is some small difference between 1.6 and 1.10 xlogs, which makes 1.6 xlogs appear erroneous to 1.10+ instances.
+In order to work around this, start the instance in ``force_recovery`` mode.
```
> In order to work around this, start the instance in ``force_recovery`` mode.

When and where a user should set ``box.cfg{force_recovery = true}`` ? In the init.lua starting file before launching the instance after upgrade? Better to make it more clear.

veod32 commented on doc/book/admin/upgrades/1.6-1.10.rst:
```diff
@@ -0,0 +1,59 @@
+    a[problematic_field_no].type = 'number'
+    box.space.problematic_space:format(a)
+
+Once this is performed on the master, it's safe to proceed to step 8, making snapshot on every node.
```
Where a user should run this piece of Lua code
- in TT console 
- or put it in the init.lua starting file before launching the instance
- or as a separate script file that will be run after performing ``box.schema.upgrade()``

Better to make it more clear.

veod32 replied:
> to proceed to step 8

Be seeing that, I started to read the paragraph below **Step 8:** The user...
instead of going to the real step in the procedure above.

veod32 commented on doc/book/admin/upgrades/1.6-2.0-downtime.rst:
```diff
@@ -0,0 +1,20 @@
+..  _admin-upgrades-1.6-2.x_downtime:
+
+Upgrade from 1.6 directly to 2.x
```
May to  rename to 

Upgrade from 1.6 directly to 2.x with downtime

to make it clear what the case is (it this is the case, of course )))

patiencedaur committed:
Add explicit box.schema.upgrade description

It's better than it used to be.
Now the Upgrades section can be restructured a bit.

patiencedaur committed:
Apply suggestions from review

* Some text is repeated. This is because Sphinx cannot correctly
  enumerate lists when list items come from different `include` files.

* Clarify the "replica" term in the instructions

* Clarify how to change the master in a replica set

* Add cross-links to guide the user

* Bring indents in accordance with the style guide

* Clarify how to run code to fix xlogs when upgrading from 1.6 to 1.10
